Big lock

jenswi-linaro opened the pull request:
These patches are needed by the generic kernel driver, but they are also applicable for or main branch as they in some sense makes the code easier. We are also moving in this direction with the multi threaded OP-TEE.

jforissier commented:
This does not seem to work too well on HiKey :(
I tried:
```
root@HiKey:/ xtest 4002 & xtest 4002 &
root@HiKey:/ 
TEE test application started with device [opteearmtz00]

TEE test application started with device [opteearmtz00]
######################################################
#
# XTEST_TEE_TEST
#
######################################################
 
* XTEST_TEE_4002 Test TEE Internal API MAC operations
######################################################
#
# XTEST_TEE_TEST
#
######################################################
 
* XTEST_TEE_4002 Test TEE Internal API MAC operations

```
Both apps are stuck in 'D' state as shown by 'top', i.e., uninterruptible sleep. Am I missing some driver updates?

jenswi-linaro commented:
Thanks, I can reproduce the problem on FVP too. I'll look into it.

jenswi-linaro commented:
The big lock was released while holding another mutex, I need to redesign the big lock scheme a bit.

jenswi-linaro commented:
Rebased and redesigned the locking.
The wait_queue commit depends on https://github.com/OP-TEE/optee_linuxdriver/pull/23
This pull request has grown a bit and is perhaps a bit too large. I'll split it if needed.

jenswi-linaro commented:
Has anyone had a chance to look at this?

jforissier commented:
Hello @jenswi-linaro, this looks OK, and I can give a:

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey ARM64)`

I have tested with `CFG_NUM_THREADS=8` in addition to the default value (2), running several instances of xtest simultaneously:
```
for i in `seq 1 4` ; do (xtest 4002 &) ; done | grep "subtests"
```
No problem detected except some out-of-memory conditions as the number of applications becomes >= 6.

jenswi-linaro commented:
This PR has two especially interesting concepts that it's important that we get right.
1. A wait queue primitive which needs a new API from the Linux driver
2. Entry of OP-TEE will only return busy when it's out of threads (a thread is required to make an RPC for wait queue sleep). This simplifies the synchronization required in the driver when entering OP-TEE, especially if the driver doesn't know whether OP-TEE can run on multiple cores in parallel or not.

jbech-linaro commented on core/arch/arm/kernel/tee_ta_manager.c:
```diff
@@ -109,6 +109,33 @@ static void set_tee_rs(struct tee_ta_session *tee_rs)
+
+		mutex_lock(&tee_ta_mutex);
+		old_busy = ctx->busy;
+		if (!old_busy)
```
The check isn't necessary since `ctx->busy` will always be set to `true` here regardless if it has value `false` or `true` when we are calling this function, right? However, by doing so it seems like you can move stuff around to get rid of `old_busy`. The `old_busy` is used outside the mutex, so it might be that I have missed something (I haven't read all changes yet, so I might revise this comment when that has been done).

jenswi-linaro replied:
Sure, we can remove the if. `old_busy` is needed since I'd like to match the `mutex_lock()` with one `mutex_unlock()` and I need to know if some other thread already holds `ctx->busy`.

jenswi-linaro commented on core/arch/arm/kernel/mutex.c:
```diff
@@ -26,212 +26,42 @@
 		DMSG("thread: %u sleeping", thread_get_id());
-		did_sleep = true;
-		mutex_wait_cmd(TEE_WAIT_MUTEX_SLEEP, handle, tick);
+		wq_wait(&m->wq);
```
There's race here, when the mutex becomes unlocked before wq_wait has added its wait element to the wait queue.

jenswi-linaro commented:
Race problem in wait_queue/mutex solved, rebased on latest.

jenswi-linaro commented:
This PR is not intended to be merged yet, it's a request for comments.

I've rebased the patches and added a condvar primitive resembling pthread_condvar

jenswi-linaro commented:
Rebased on master. Please review.

pascal-brand-st-dev commented:
Review is quite positive. Code is clear, and nothing important has been found. Note that if 5 commits will be finally merged instead of a big 1, then the commit "add condvar synchro primitive" should come earlier.

Also, I think that a .md doc file would be really useful.

jenswi-linaro commented:
Thanks, I'll add an .md file.
Here in the PR the order of the commits doesn't reflect the real order of the commits. I would prefer to keep each commit separated so when we merge this PR it will be as five (or perhaps six with an .md file also) separate commits.

jenswi-linaro commented:
Added some documentation on synchronization.

jenswi-linaro commented:
Rebased, depends on https://github.com/OP-TEE/optee_test/pull/28 to pass xtest due to the "core: avoid deadlocks caused by single-instance" patch.

jenswi-linaro commented:
Has anyone had a chance to look at this?

jforissier commented on core/arch/arm/kernel/mutex.c:
```diff
@@ -26,187 +26,71 @@
+
+		/*
+		 * If the mutex is locked we need to initialize the wqe
+		 * before releasing the spinlock to guerantee that we don't
```
guarantee

jforissier commented on core/arch/arm/kernel/tee_ta_manager.c:
```diff
@@ -115,6 +116,55 @@ static void set_tee_rs(struct tee_ta_session *tee_rs)
+	 * If we're about to set a context of a single instance TA busy
+	 * make sure that we don't have any other single instance TA in the
+	 * call stack to guarantee that we can't deadlock.
+	 */
```
Isn't this a bit too strict?

jenswi-linaro replied:
I'm open for suggestions. It has to guarantee that we can't deadlock and it would be an advantage if TAs see a consistent behavior.

jforissier replied:
I was thinking about actual deadlock detection, something like: if `ctx->busy` is true, and `ctx` is single instance TA, and `ctx` is already waiting for some single instance TA that we have in the current stack, then we are about to deadlock, otherwise it's OK.
Now, assuming the above condition is correct, I'm not sure it is worth the trouble (and the associated overhead).

jenswi-linaro replied:
What you're describing isn't enough to detect every possible deadlock, what if it's a more complicated lock graph involving multiple concurrent threads?

Agree, let's keep it simple for now.

jforissier commented on core/arch/arm/kernel/wait_queue.c:
```diff
@@ -0,0 +1,216 @@
+	struct teesmc32_param params[2];
+
+	DMSG("%s thread %u",
+	     cmd == TEE_RPC_WAIT_QUEUE_SLEEP ? "sleep" : "wake ", id);
```
Extra space after "wake " (and same below). Define this expression as a macro maybe?

jenswi-linaro replied:
The extra space is intentional for the log entries to line up better. It can be achieved with `"%-5s"` or something instead if you prefer that.

jforissier replied:
I see, that's fine.

jforissier commented on core/arch/arm/kernel/wait_queue.c:
```diff
@@ -0,0 +1,216 @@
+		tee_ta_set_current_session(sess);
+}
+
+static void slist_add_tail(struct wait_queue *wq, struct wait_queue_elem *wqe)
```
By making `wait_queue` a `SIMPLEQ` rather than a `SLIST`, you could transform this into a constant-time operation. Wouldn't it be better?

jenswi-linaro replied:
Worst case is the number of thread which always will be a low number. Even if it would be a high number it's still only slow path processing, fast path is not affected by this. By using a `SLIST` instead of a `SIMPLEQ` we save 8 or 4 bytes in each mutex.

jforissier replied:
OK.

jforissier commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
 therefore this is something that needs to be implemented by each platform.
 
+## Synchronization
+OP-TEE has three primitives for synchornization of threads and CPUs,
```
synchronisation

pascal-brand-st-dev replied:
s/CPUs,/CPUs:

jforissier commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+A condvar is represented by `struct condvar`. A condvar is similar to a
+pthread_condvar_t in the pthreads standard, only less advanced. Condition
+variables are used to wait for some condition to be fulfilled and are
+always used together a mutex. Once a condition variable has been used with
```
together with

jforissier commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+
+A condvar is initialized with `CONDVAR_INITIALIZER` or `condvar_init()`.
+
+`condvar_wait()` atomically unlocks the supplied mutex waits in normal
```
and waits

jforissier commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+function returns the mutex is locked again.
+
+`condvar_signal()` wakes up one waiter of the condition variable (waiting
+in `condvar_wait()`
```
missing `)`

pascal-brand-st-dev commented on core/arch/arm/kernel/thread.c:
```diff
@@ -1038,6 +1014,58 @@ void thread_kernel_disable_vfp(uint32_t state)
+	bool have_bl = false;
+	bool have_other = false;
+
+	TAILQ_FOREACH(m, &threads[ct].mutexes, link) {
```
We're only checking mutexes. Should we check spin-lock and condvar also?

jenswi-linaro replied:
No, only mutexes here.

pascal-brand-st-dev commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+`cpu_spin_lock()` locks a spin-lock
+
+`cpu_spin_trylock()` locks a spin-lock if unlocked and returns `0` else if
+the spin-spin else the spin-lock is unchanged and the function returns `!0`.
```
I do not understand "else if the spin-spin else"

jenswi-linaro replied:
Me neither. :-) I'll remove some words.

pascal-brand-st-dev commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+The *big lock* mutex protects secure world against unsupported concurrent
+execution. The *big lock* is taken just before the SMC handler is called
+and released when the function has returned. The thread unlocks the *big
+lock* lock when concurrency can be enabled. The *big lock* should be locked
```
"should be locked" or "should be unlocked".

jenswi-linaro replied:
I don't understand the comment, but I'll rephrase to make the text clearer.

pascal-brand-st-dev commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+doesn't miss the signal.
+
+## Big lock
+The *big lock* mutex protects secure world against unsupported concurrent
```
Would be worth to know what is "unsupported". Right now, we have only 1 thread active in the trusted os. But when leaving the trusted os (rpc or normal exit), then we're able to take the execution of another TA.

jenswi-linaro replied:
Rephrased this.

jenswi-linaro commented:
Addressed all review comments so far.

jbech-linaro commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,90 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+`mutex_destroy()` asserts that the mutex is unlocked and there's no
+waiters, after this the memory used by the mutex can be freed.
+
+When a mutex is locked it's own by the thread calling `mutex_lock()` or
```
owned

jbech-linaro commented on documentation/optee_design.md:
```diff
@@ -198,6 +198,90 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
+function returns the mutex is locked again.
+
+`condvar_signal()` wakes up one waiter of the condition variable (waiting
+in `condvar_wait()`)
```
In the rendered version I cannot see the ending parenthesis ... strange.

jenswi-linaro replied:
I can see it, I guess you need to reload or something.

pascal-brand-st-dev commented:
Without the commit "core: avoid deadlocks caused by single-instance TA":

    Tested-by: Pascal Brand <pascal.brand@linaro.org> (QEMU platform)
    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

jforissier commented:
Like @pascal-brand-st-dev I have tested this PR after reverting commit "core: avoid deadlocks caused by single-instance TA", which causes a few failures in the GP test suite (extended xtest).

In this configuration:
```
Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>
Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)
```

jenswi-linaro commented:
Removed the commit "core: avoid deadlocks caused by single-instance TA"
Rebased and tags applied
